<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="author" content="Zhili">
<meta name="description" content="Sed magna purus, fermentum eu, tincidunt eu, varius ut, felis. Morbi mollis tellus ac sapien. Vestibulum ullamcorper mauris at ligula. In hac habitasse platea dictumst. Vivamus in erat ut urna cursus vestibulum.
    ">
<meta name="keywords" content="博客, 技术, 生物信息，精准医疗, 遗传咨询, 高通量技术, 算法, 学习, 作品, 写作, C&#43;&#43;, Perl">
<meta name="referrer" content="always">
<title>[MECpp]Item-13 Catch Exceptions by Reference - sy123.ml</title>
<link rel="stylesheet" type="text/css" href="/css/main.css">
<link rel="icon" href="/logo/favicon.ico" type="image/x-icon">
</head>
<body>
<header>
  <div>
    <h1><a href="https://sy123.ml/">sy123.ml</a></h1><h2><a href="https://sy123.ml/">winter is coming.</a></h2>
  </div>
  <nav><a href="/tags/">Tags</a><a href="/about/">About</a></nav>
</header>
<main>
  <article>
    <div class="title">
  <h1>[MECpp]Item-13 Catch Exceptions by Reference</h1>
  </div>
<div class="meta">
  <div>2018-04-05 18:57</div>
  <div>
    <span><a href="https://sy123.ml/tags/technique/">#technique</a></span>
      <span><a href="https://sy123.ml/tags/cpp/">#cpp</a></span>
      </div>
  </div>
<div class="content">
  <p>If catching by reference, we sidestep the questions about object deletion, avoid slicing exception objects, retain the ability to catch standard exceptions, and limit the number of times exception objects being copied.</p>
<!-- toc -->
<h1 id="catch-by-pointer">Catch by pointer</h1>
<p>In order to catch by pointer, programmers need to define exception objects in a way that <strong>guarantees the objects exist</strong> after control leaves the <code>throw</code> site. Global and static objects work fine, but it's easy for programmers to forget the constraint:</p>
<div class="highlight"><div style="color:#272822;background-color:#fafafa;-moz-tab-size:4;-o-tab-size:4;tab-size:4">
<table style="border-spacing:0;padding:0;margin:0;border:0;width:auto;overflow:auto;display:block;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre style="color:#272822;background-color:#fafafa;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">1
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">2
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">3
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">4
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">5
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">6
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre style="color:#272822;background-color:#fafafa;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-cpp" data-lang="cpp"><span style="color:#00a8c8">void</span> <span style="color:#75af00">someFunction</span><span style="color:#111">(</span><span style="color:#111">)</span>
<span style="color:#111">{</span>
    <span style="color:#111">exception</span> <span style="color:#111">ex</span><span style="color:#111">;</span>
    <span style="color:#111">.</span><span style="color:#111">.</span><span style="color:#111">.</span>
    <span style="color:#00a8c8">throw</span> <span style="color:#f92672">&amp;</span><span style="color:#111">ex</span><span style="color:#111">;</span>   <span style="color:#75715e">// throw a pointer to an object 
</span><span style="color:#75715e"></span>    <span style="color:#111">.</span><span style="color:#111">.</span><span style="color:#111">.</span>          <span style="color:#75715e">// that&#39;s about to be destroyed
</span><span style="color:#75715e"></span><span style="color:#111">}</span>
</code></pre></td></tr></table>
</div>
</div><p>An alternative is to throw a pointer to a new heap object:</p>
<div class="highlight"><div style="color:#272822;background-color:#fafafa;-moz-tab-size:4;-o-tab-size:4;tab-size:4">
<table style="border-spacing:0;padding:0;margin:0;border:0;width:auto;overflow:auto;display:block;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre style="color:#272822;background-color:#fafafa;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">1
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">2
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">3
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">4
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">5
</span><span style="margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#7f7f7f">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre style="color:#272822;background-color:#fafafa;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-cpp" data-lang="cpp"><span style="color:#00a8c8">void</span> <span style="color:#75af00">someFunction</span><span style="color:#111">(</span><span style="color:#111">)</span>
<span style="color:#111">{</span>
    <span style="color:#111">.</span><span style="color:#111">.</span><span style="color:#111">.</span>
    <span style="color:#00a8c8">throw</span> <span style="color:#00a8c8">new</span> <span style="color:#111">exception</span><span style="color:#111">;</span> <span style="color:#75715e">// throw a pointer to a new heap-based object 
</span><span style="color:#75715e"></span>    <span style="color:#111">.</span><span style="color:#111">.</span><span style="color:#111">.</span>  <span style="color:#75715e">// hopefully the objerator new will not itself throw any exception
</span><span style="color:#75715e"></span><span style="color:#111">}</span>
</code></pre></td></tr></table>
</div>
</div><p>This design will make a hard time for authors of <code>catch</code> clauses: <strong>to delete or not to delete</strong> the pointer they receive? In the <code>catch</code> site, we can't tell if an exception object is allocated on the heap or defined as a global (or static) object.</p>
<p>Furthermore, catch-by-pointer runs <strong>contrary to the convention</strong> of the language: the four standard exception<sup id="fnref:1"><a href="#fn:1" class="footnote-ref" role="doc-noteref">1</a></sup> - <code>bad_alloc</code>, <code>bad_cast</code>, <code>bad_typeid</code>, and <code>bad_exception</code> are all objects, rather than pointers to objefts, so we have to catch them by value or by reference.</p>
<h1 id="catch-by-value">Catch by value</h1>
<p>Catch-by-value requires exception objects be copied <em>twice</em> each time they thrown (MECpp item 12), and it also has <strong><em>slicing problem</em></strong>: derived class exception objects caught as base class exceptions have their derivedness &ldquo;sliced off&rdquo; : they lack derived class data members, and resolve to base class virtual functions (the same behavior as when an object is passed to a function by value)</p>
<h1 id="catch-by-reference">Catch by reference</h1>
<p>Catch-by-reference suffers from none of the problems:</p>
<ul>
<li>Unlike catch-by-pointer, the question of object deletion fails to arise, and there's no difficulty in catching the standard exception types</li>
<li>Unlike catch-by-value, there is no slicing problem, and exception objects are copied only once.</li>
</ul>
<p>Of course, if there's no need to change the exception object in the <code>catch</code> site, we'd catch not just reference, but by reference to <code>const</code>.</p>
<section class="footnotes" role="doc-endnotes">
<hr>
<ol>
<li id="fn:1" role="doc-endnote">
<p><code>bad_alloc</code>: thrown when <code>operator new</code> can't satisfy a memory request (MECpp item 8); <code>bad_cast</code>: thrown when a <code>dynamic_cast</code> to a reference fails (MECpp item 2); <code>bad_typeid</code>: thrown when <code>typeid</code> is applied to a dereferenced null pointer; and <code>bad_exception</code>: available for unexpected exceptions (MECpp item 14) <a href="#fnref:1" class="footnote-backref" role="doc-backlink">&#x21a9;&#xfe0e;</a></p>
</li>
</ol>
</section></div>

  </article>
</main>
<footer>
  <div>
    <div>
      2007-2020 Zhili.</div>
  </div>
</footer>
</body>
</html>
